File: LargestRectangleinHistogram.java
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack = []
        maxArea = 0
        for i in range(len(heights)):
            index = i
            while (stack and stack[-1][1] > heights[i]):
                index = stack[-1][0]
                maxArea = max(maxArea, (i-index)*stack[-1][1])
                stack.pop()
            stack.append((index, heights[i]))
        for i in range(len(stack)):
            maxArea = max(maxArea, (len(heights)-stack[-1][0])*stack[-1][1])
            stack.pop()
        return maxArea
        

File: LargestRectangleinHistogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        maxArea = 0
        for i in range(len(heights)):
            start = i
            while(stack and heights[i] < stack[-1][1]):
                val = stack.pop()
                maxArea = max(maxArea, (i-val[0])*val[1])
                start = val[0]
            stack.append((start, heights[i]))
        while(stack):
            lastval = stack.pop()
            maxArea = max(maxArea, (len(heights)-lastval[0])*lastval[1])
        return maxArea
